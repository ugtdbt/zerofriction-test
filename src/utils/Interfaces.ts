export interface State<T> {
  data: T;
  valideAddressForm: boolean;
  valideContactDetailsForm: boolean;
  valideOrganizationDetailsForm: boolean;
  success: boolean;
  save: boolean;
  cancel: boolean;
  formtouched: boolean;
  formreset: boolean;
}

export interface Action<T> {
  type: string;
  payload?: T;
}
