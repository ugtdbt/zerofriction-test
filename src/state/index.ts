export * as actionCreators from "./actionCreators/organisationConfigurationActionCreator";
export * from "./store";
export * from "./reducers/index";
