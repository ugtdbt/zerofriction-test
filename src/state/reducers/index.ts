import { combineReducers } from "redux";
import organisationConfigurationReducer from "./organisationConfigurationReducer";

const reducers = combineReducers({
  organisationConfiguration: organisationConfigurationReducer,
});

export default reducers;

export type State = ReturnType<typeof reducers>;
