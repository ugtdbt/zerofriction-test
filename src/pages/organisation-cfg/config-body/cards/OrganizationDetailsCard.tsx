import React, { memo, useState, useEffect } from "react";
import { Card, Form, Checkbox, Input } from "antd";
import type { State } from "../../../../state";
import { useSelector, useDispatch } from "react-redux";
import type { OrganizationConfigType } from "../../../../utils/types/organisation-config";
import type { FormInstance } from "antd/lib/form";
import { bindActionCreators } from "redux";
import { actionCreators } from "../../../../state";
import type { CheckboxChangeEvent } from "antd/lib/checkbox";

const OrganizationDetailsCard: React.FC = memo(() => {
  const initialValidationData = {
    code: {
      validateStatus: "",
      help: "",
    },
    description: {
      validateStatus: "",
      help: "",
    },
    bankAccount: {
      validateStatus: "",
      help: "",
    },
    vatAccountNumber: {
      validateStatus: "",
      help: "",
    },
    companyAccountNumber: {
      validateStatus: "",
      help: "",
    },
  };

  const formRef = React.useRef<FormInstance>(null);
  const { data, formreset, save } = useSelector(
    (state: State) => state.organisationConfiguration
  );

  useEffect(() => {
    formRef.current?.resetFields();
    setFormErrors(initialValidationData);
    setFormDisabled(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [formreset]);

  const [formValues, setFormValues] = useState<OrganizationConfigType>(
    data.organizationConfig
  );

  const [formDisabled, setFormDisabled] = useState<boolean>(false);

  const [formErrors, setFormErrors] = useState(initialValidationData);

  const handleAddressChange = (e: any) => {
    const { id, value } = e.target;
    setFormValues({ ...formValues, [id]: value });
    validate({ ...formValues, [id]: value }, id);
    validateForm({ ...formValues, [id]: value });
  };

  const dispatch = useDispatch();
  const { valideOrganizationDetails, clickOrganizationDetailsSaveButton } =
    bindActionCreators(actionCreators, dispatch);

  const validateForm = (formValues: OrganizationConfigType) => {
    let fromValide = true;
    if (formValues["code"] === "") {
      fromValide = false;
    }
    if (formValues["description"] === "") {
      fromValide = false;
    }
    if (formValues["bankAccount"] === "") {
      fromValide = false;
    }
    if (formValues["vatAccountNumber"] === "") {
      fromValide = false;
    }
    if (formValues["companyAccountNumber"] === "") {
      fromValide = false;
    }
    valideOrganizationDetails(fromValide);
  };

  const validate = (formValues: OrganizationConfigType, key: string) => {
    switch (key) {
      case "code":
        if (formValues[key] === "") {
          setFormErrors({
            ...formErrors,
            code: {
              validateStatus: "error",
              help: "code is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            code: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "description":
        if (formValues[key] === "") {
          setFormErrors({
            ...formErrors,
            description: {
              validateStatus: "error",
              help: "description is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            description: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "bankAccount":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            bankAccount: {
              validateStatus: "error",
              help: "bankAccount is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            bankAccount: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "vatAccountNumber":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            vatAccountNumber: {
              validateStatus: "error",
              help: "vatAccountNumber is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            vatAccountNumber: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "companyAccountNumber":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            companyAccountNumber: {
              validateStatus: "error",
              help: "company Account Number is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            companyAccountNumber: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      default:
        break;
    }
  };

  const handleMigrationModeChange = (e: CheckboxChangeEvent) => {
    setFormValues({ ...formValues, migrationMode: e.target.checked });
  };

  useEffect(() => {
    if (save) {
      clickOrganizationDetailsSaveButton(formValues);
      setFormDisabled(true);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [save]);

  return (
    <Card title="Organization Details" data-testid="organization-details-form">
      <Form
        labelCol={{ span: 10 }}
        wrapperCol={{ span: 14 }}
        layout="horizontal"
        ref={formRef}
        disabled={formDisabled}
      >
        <Form.Item
          label="Migration Mode"
          name="migrationMode"
          valuePropName="checked"
        >
          <Checkbox onChange={handleMigrationModeChange}></Checkbox>
        </Form.Item>

        <Form.Item
          label="code"
          name="code"
          validateStatus={
            formErrors.code.validateStatus === "error" ? "error" : undefined
          }
          help={formErrors.code.help !== "" ? formErrors.code.help : undefined}
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>

        <Form.Item
          label="description"
          name="description"
          validateStatus={
            formErrors.description.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.description.help !== ""
              ? formErrors.description.help
              : undefined
          }
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>

        <Form.Item
          label="bankAccount"
          name="bankAccount"
          validateStatus={
            formErrors.bankAccount.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.bankAccount.help !== ""
              ? formErrors.bankAccount.help
              : undefined
          }
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>
        <Form.Item
          label="vatAccountNumber"
          name="vatAccountNumber"
          validateStatus={
            formErrors.vatAccountNumber.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.vatAccountNumber.help !== ""
              ? formErrors.vatAccountNumber.help
              : undefined
          }
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>
        <Form.Item
          label="companyAccountNumber"
          name="companyAccountNumber"
          validateStatus={
            formErrors.companyAccountNumber.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.companyAccountNumber.help !== ""
              ? formErrors.companyAccountNumber.help
              : undefined
          }
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>
      </Form>
    </Card>
  );
});

OrganizationDetailsCard.displayName = "OrganizationDetailsCard";
export default OrganizationDetailsCard;
