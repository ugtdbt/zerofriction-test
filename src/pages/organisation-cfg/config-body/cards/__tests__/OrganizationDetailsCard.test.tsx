import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import OrganizationDetailsCard from "../OrganizationDetailsCard";
import { store } from "../../../../../state/store";
import { Provider } from "react-redux";
import "../../../../../utils/matchMedia";

describe("AddressCard component", () => {
  const OrganizationDetailsCardComponent = (
    <React.StrictMode>
      <Provider store={store}>
        <OrganizationDetailsCard />
      </Provider>
    </React.StrictMode>
  );

  afterEach(() => {
    cleanup();
  });

  it("OrganizationDetailsCardComponent Match Snapshot", () => {
    const tree = renderer.create(OrganizationDetailsCardComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render OrganizationDetailsCardComponent component", () => {
    render(OrganizationDetailsCardComponent);
    const mainElem = screen.getByTestId("organization-details-form");
    expect(mainElem).toBeInTheDocument();
  });
});
