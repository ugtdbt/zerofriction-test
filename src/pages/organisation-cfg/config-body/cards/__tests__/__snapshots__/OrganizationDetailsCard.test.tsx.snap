// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressCard component OrganizationDetailsCardComponent Match Snapshot 1`] = `
<div
  className="ant-card ant-card-bordered css-dev-only-do-not-override-1km3mtt"
  data-testid="organization-details-form"
>
  <div
    className="ant-card-head"
    style={Object {}}
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        Organization Details
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <form
      className="ant-form ant-form-horizontal css-dev-only-do-not-override-1km3mtt"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="migrationMode"
              title="Migration Mode"
            >
              Migration Mode
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <label
                  className="ant-checkbox-wrapper ant-checkbox-wrapper-in-form-item css-dev-only-do-not-override-1km3mtt"
                >
                  <span
                    className="ant-checkbox css-dev-only-do-not-override-1km3mtt"
                    style={Object {}}
                  >
                    <input
                      checked={false}
                      className="ant-checkbox-input"
                      disabled={false}
                      id="migrationMode"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyPress={[Function]}
                      onKeyUp={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="ant-checkbox-inner"
                    />
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="code"
              title="code"
            >
              code
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="code"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="description"
              title="description"
            >
              description
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="description"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="bankAccount"
              title="bankAccount"
            >
              bankAccount
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="bankAccount"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="vatAccountNumber"
              title="vatAccountNumber"
            >
              vatAccountNumber
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="vatAccountNumber"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="companyAccountNumber"
              title="companyAccountNumber"
            >
              companyAccountNumber
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="companyAccountNumber"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
